.adr/0017-missed-chain-events.md

```diff
@@ -0,0 +1,131 @@
+# 0017 Missed Chain Events
+
+## Status
+
+Review
+
+## Context
+
+It is critical that a go-nitro node can synchronise pertinent blockchain state, in order to successfully perform off chain protocols and protect against faulty challenges. For efficiency reasons, a node synchronises that state by listening to blockchain events. Missing those events can therefore cause the node's data to get out of sync with other nodes and the chain state. Events might be missed if the node has not been continuously running throughout the lifetime of any particular channel it participates in. For example, there may have been a period of downtime or loss of internet due to power loss or maintenance.
+
+To address these concerns, a nitro node should check for any chain events it might have missed since it was last online, and queue those events to be processed as part of the node's initialization routines.
+
+```go
+	ecs.wg.Add(3)
+	go ecs.listenForEventLogs(errChan, eventSub, eventChan, eventQuery)
+	go ecs.listenForNewBlocks(errChan, newBlockSub, newBlockChan)
+	go ecs.listenForErrors(errChan)
+
+	// Search for any missed events emitted while this node was offline
+	err = ecs.checkForMissedEvents(startBlock)
+	if err != nil {
+		return nil, err
+	}
+```
+
+There are a few decisions to make within this design:
+
+1. How does the node protect against race conditions that could caused missed chain events during initialization?
+2. How does the node ensure it processes chain events idempotently (i.e. how does it ensure the same chain event does not make changes to a `Channel`'s data multiple times?). This situation could occur if the node restarts, scans through blocks for events during initialization, and detects events that it has already processed.
+3. When/how should the `lastBlockNum` be updated in the `store`? When the node is initialized it can read the `lastBlockNumSeen` from the `store`, and search for any chain events that occurred between that block and the current block. If the node uses a `memstore`, then the `lastBlockNumSeen` will always be set to `0` when the node is first initialized. By updating the `lastBlockNum` in `store`, the `chainservice` will not have to scan through as many old blocks when it first initializes and calls `checkForMissedEvents`.
```

Comment:
```suggestion
3. By storing a `lastBlockNum` variable in the `store`,  the `chainservice` will not have to scan through as many old blocks when it first initializes and calls `checkForMissedEvents`. When the node is initialized it can read the `lastBlockNumSeen` from the `store`, and search for any chain events that occurred between that block and the current block. If the node uses a `memstore`, then the `lastBlockNumSeen` will always be set to `0` when the node is first initialized. When/how should the `lastBlockNumSeen` be updated?
```

Reply:
Change made in https://github.com/statechannels/go-nitro/pull/1624/commits/06b800ff68d376a29ee776e21e9b8a074fcd8ff6